WalletWasabi/WabiSabi/Backend/Banning/CoinVerifierApiClient.cs

```diff
@@ -1,61 +1,61 @@
 using NBitcoin;
 using Newtonsoft.Json;
-using System.Collections.Generic;
-using System.Linq;
 using System.Net;
 using System.Net.Http;
-using System.Runtime.CompilerServices;
 using System.Threading;
 using System.Threading.Tasks;
 using WalletWasabi.Logging;
 using WalletWasabi.WabiSabi.Backend.Statistics;
 
 namespace WalletWasabi.WabiSabi.Backend.Banning;
 
-public class CoinVerifierApiClient
+public class CoinVerifierApiClient : IAsyncDisposable
 {
-	public CoinVerifierApiClient(string token, Network network, HttpClient httpClient)
-	{
-		ApiToken = token;
-		Network = network;
-		HttpClient = httpClient;
-	}
-
-	public CoinVerifierApiClient() : this("", Network.Main, new() { BaseAddress = new("https://www.test.test") })
-	{
-	}
+	/// <summary>Maximum number of actual HTTP requests that might be served concurrently by the CoinVerifier webserver.</summary>
+	public const int MaxParallelRequestCount = 30;
 
-	private TimeSpan TotalApiRequestTimeout { get; } = TimeSpan.FromMinutes(3);
+	/// <summary>Maximum re-tries for a single API request.</summary>
+	private const int MaxRetries = 3;
 
-	private string ApiToken { get; set; }
-	private Network Network { get; set; }
+	/// <summary>Timeout for a single <see cref="SendRequestAsync(Script, CancellationToken)"/> invocation (including retries).</summary>
+	private static TimeSpan TotalApiRequestTimeout { get; } = TimeSpan.FromMinutes(3);
 
-	private HttpClient HttpClient { get; set; }
-
-	public virtual async Task<ApiResponseItem> SendRequestAsync(Script script, CancellationToken cancellationToken)
+	public CoinVerifierApiClient(string apiToken, HttpClient httpClient)
 	{
+		ApiToken = apiToken;
+		HttpClient = httpClient;
+
 		if (HttpClient.BaseAddress is null)
 		{
 			throw new HttpRequestException($"{nameof(HttpClient.BaseAddress)} was null.");
 		}
+
 		if (HttpClient.BaseAddress.Scheme != "https")
 		{
 			throw new HttpRequestException($"The connection to the API is not safe. Expected https but was {HttpClient.BaseAddress.Scheme}.");
 		}
+	}
 
-		var address = script.GetDestinationAddress(Network.Main); // API provider don't accept testnet/regtest addresses.
+	private string ApiToken { get; }
+
+	private HttpClient HttpClient { get; }
+
+	private SemaphoreSlim ThrottlingSemaphore { get; } = new(initialCount: MaxParallelRequestCount);
+
+	public virtual async Task<ApiResponseItem> SendRequestAsync(Script script, CancellationToken cancellationToken)
+	{
+		var address = script.GetDestinationAddress(Network.Main); // API provider doesn't accept testnet/regtest addresses.
 
 		using CancellationTokenSource timeoutTokenSource = new(TotalApiRequestTimeout);
 		using CancellationTokenSource linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutTokenSource.Token);
```

Comment:
Should we perhaps move these CTSources inside the `ThrottlingSemaphore` block?

For example if 50 scripts needs to be send and the first 30 takes too much time, the last 20 script will have much less time and the cancellation token can easily kick in.

WDYT?

Reply:
What I meant is that the 31th request in parallel will wait at this line (LOC58 in `CoinVerifierApiClient`)

```cs
await ThrottlingSemaphore.WaitAsync(linkedTokenSource.Token).ConfigureAwait(false);
```

while its `timeoutTokenSource` is ticking.
So the 31th request will only have (let's say) 1 minute to get a response bc it was waiting on that line.

Reply:
> Maybe it would be for the best if you created a new PR that just removes this TotalApiRequestTimeout and the cts with the link to this discussion.

Yes, I will do that. And IMO you are right about that the caller should specify an absolute maximum time for the request + response.

> Also we should have some statista data for this to see how long these requests take.

We do have some statistic data about these request.


```cs
var before = DateTimeOffset.UtcNow;

response = await HttpClient.SendAsync(content, linkedTokenSource.Token).ConfigureAwait(false);

var duration = DateTimeOffset.UtcNow - before;
RequestTimeStatista.Instance.Add("verifier-request", duration);
```

It's not that much, but we can read the smallest-largest-average time for the requests from the logs.